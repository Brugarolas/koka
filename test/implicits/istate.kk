// --------------------------------------------------------------
// Local mutable state defined as implicit control.
// This is a much nicer way to give semantics to state
// versus the state-monad encoding. However, implementations
// need to support frame "smashing" to implement this in constant
// stack space. Paper is forthcoming...
// --------------------------------------------------------------

implicit<s> val get : s
implicit<s> fun set(x : s) : ()

fun local( init, action ) { // : (a,() -> <set<a>,get<a>|e> b) -> e b
  with val get = init
  with control set(x){
    with val get = x
    // change effect of `resume(())` from `<get<a>|e>` to `<get<a>,get<c>|e>`.
    // this has no runtime effect but is needed for type inference.
    inject behind<get>{resume(())}
  }
  action()
}

fun printn() {
  while{ get > 0 } {
    println((get:int))
    set(get - 1)
  }
}

fun main() {
  with local(5)
  printn()
}
