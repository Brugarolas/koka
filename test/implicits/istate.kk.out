[01;30moperator branch (local) .val-get: resume tail
[00;39;49m[00m[01;30moperator branch (local) .val-get: resume tail
[00;39;49m[00m[01;30moperator branch (local) set: resume normal
[00;39;49m[00m5
4
3
2
1
.Op-set : forall<a> (x : a) -> .op-set<a>
.Op-val-get : forall<a> .op-val-get<a>
.Ops-set : forall<a,b> (set : .op-set<a>) -> .ops-set<a,()>
.Ops-val-get: forall<a,b> (.op-val-get<a>) -> .ops-get<a,a>
.tag-get : string
.tag-set : string
.val-get : forall<a> () -> (get<a>) a
get : forall<a> value<a>
local : forall<a,b,e> (init : a, action : () -> <get<a>,set<a>|e> b) -> e b
main : () -> <div,console> ()
printn : () -> <get<int>,set<int>,div,console> ()
set : forall<a> (x : a) -> (set<a>) ()